src/broker/Manager.cc

```diff
@@ -92,6 +92,142 @@ void print_escaped(std::string& buf, std::string_view str) {
+        // continues.
+        uint64_t overflows = 0;
+
+        // When we last started a stats-tracking interval for this peering.
+        double last_interval = 0;
+    };
+
+    // For per-peering tracking, map endpoint IDs to the above state.
+    using EndpointMetricMap = std::unordered_map<broker::endpoint_id, Stats>;
+
+    PeerBufferState(size_t a_buffer_size, double a_stats_interval)
+        : buffer_size(a_buffer_size), stats_interval(a_stats_interval) {
+        stats_table =
+            zeek::make_intrusive<zeek::TableVal>(zeek::id::find_type<zeek::TableType>("BrokerPeeringStatsTable"));
+        stats_record_type = zeek::id::find_type<zeek::RecordType>("BrokerPeeringStats");
+    }
+
+    // Update the peering's stats. This runs in Broker's execution context.
+    // Broker does not expose send-buffer/queue state explicitly, so track
+    // arrivals (a push, push_or_pull == true) and departures (pull,
+    // push_or_pull == false) as they happen.
+    void Observe(const broker::endpoint_id& peer, bool push_or_pull) {
+        std::lock_guard<std::mutex> lock(mutex);
+        auto it = stats_map.find(peer);
+
+        if ( it == stats_map.end() ) {
+            Stats stats{peer};
+            stats_map.emplace(peer, stats);
+            stats_table->Assign(stats.peer_id, zeek::make_intrusive<zeek::RecordVal>(stats_record_type));
+            it = stats_map.find(peer);
+        }
+
+        auto& stats = it->second;
+        auto now = zeek::util::current_time();
+
+        if ( now - stats.last_interval > stats_interval ) {
+            stats.last_interval = now;
+            stats.max_queued_recently = 0;
+        }
+
+        if ( stats.queued == 0 ) {
+            // Watch for underflows. We could report somehow. Note that this
+            // runs in the context of Broker's threads.
+            assert(push_or_pull);
+        }
+
+        if ( push_or_pull && stats.queued == buffer_size )
+            stats.overflows += 1;
+        else {
+            stats.queued += push_or_pull ? 1 : -1;
+            if ( stats.queued > stats.max_queued_recently )
+                stats.max_queued_recently = stats.queued;
+        }
+    }
+
+    // Updates the internal table[string] of BrokerPeeringStats and returns it.
+    const zeek::TableValPtr& GetPeeringStatsTable() {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        for ( auto& [peer, stats] : stats_map ) {
+            auto stats_v = stats_table->Find(stats.peer_id);
+
+            assert(stats_v != nullptr);
+
+            // We may get here more than stats_interval after the last
+            // Observe(), in which case the max_queued_recently value is now
+            // stale. Reset if so.
+            auto now = zeek::util::current_time();
+
+            if ( now - stats.last_interval > stats_interval ) {
+                stats.last_interval = now;
+                stats.max_queued_recently = 0;
```

Comment:
I guess it matters more here. It'd be weird to observe a record with `queued > 0` but `max_queued_recently == 0,`, no?

Reply:
Agree. Will tweak.